lib/modules/helpers.js

```diff
@@ -1,9 +1,12 @@
-function generateValidLink(url, username = '') {
+function generateValidLink(url, username) {
   let validLink = url
-  const validRegex = /^https?:\/\/(?:www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b(?:[-a-zA-Z0-9()@:%_\+.~#?&\/=]*)$/
+  let validUsername = username || 'all-contributors'
 
-  if (!validLink.startsWith("http")) validLink = `http://${url}/`
-  if (!validLink.match(validRegex)) validLink = `https://github.com/${username}/`
+  const validRegex = /^(http:\/\/www\.|https:\/\/www\.|http:\/\/|https:\/\/)?[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\.[a-z]{2,5}(:[0-9]{1,5})?(\/.*)?$/
+  const validRegexWithoutProtocol = /^[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\.[a-z]{2,5}(:[0-9]{1,5})?(\/.*)?$/
+
+  if (validLink.match(validRegexWithoutProtocol)) validLink = `http://${url}/`
+  if (!validLink.match(validRegex)) validLink = `https://github.com/${validUsername}/`
```

Comment:
One suggestion here, this is basically creating `html_url` that was already in the calling function. If how we call this function is changed like

```js
getUserLink(blog, html_url)
```

This function can be changed into

```js
function getUserLink(blog, githubProfileURL) {
	const validRegexWithScheme = /same as now/
	const validRegexWithoutScheme = /same as now/
	if (validRegexWithProtocol.test(blog)) return blog;
	if (validRegexWithoutProtocol.test(blog)) return `http://${blog}`;
	return githubProfileURL
}

```
This way, someone reading the code in the future only have to think about which `return` would be triggered. 
What do you think?

Reply:
@Roshanjossey I somehow agree however we dp have tests that somehow act as the documentation for that method. I do not think this is necessary to add more lines. I like to simplify things rather than make them complicated.

Reply:
I'll rephrase what I commented earlier. I'm suggesting to change

```js
function generateValidLink(url, username) {
  let validLink = url
  let validUsername = username || 'all-contributors'

  const validRegex = /^(http:\/\/www\.|https:\/\/www\.|http:\/\/|https:\/\/)?[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\.[a-z]{2,5}(:[0-9]{1,5})?(\/.*)?$/
  const validRegexWithoutProtocol = /^[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\.[a-z]{2,5}(:[0-9]{1,5})?(\/.*)?$/

  if (validLink.match(validRegexWithoutProtocol)) validLink = `http://${url}/`
  if (!validLink.match(validRegex)) validLink = `https://github.com/${validUsername}/`

  return validLink
}
```
 to
 
 ```js
 function getUserLink(blog, githubProfileURL) {
	const validRegexWithScheme = /same as now/
	const validRegexWithoutScheme = /same as now/
	if (validRegexWithSchema.test(blog)) return blog;
	if (validRegexWithoutSchema.test(blog)) return `http://${blog}`;
	return githubProfileURL
}
 ```
 Isn't this simpler?

Reply:
> I'll rephrase what I commented earlier. I'm suggesting to change
> 
> ```js
> function generateValidLink(url, username) {
>   let validLink = url
>   let validUsername = username || 'all-contributors'
> 
>   const validRegex = /^(http:\/\/www\.|https:\/\/www\.|http:\/\/|https:\/\/)?[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\.[a-z]{2,5}(:[0-9]{1,5})?(\/.*)?$/
>   const validRegexWithoutProtocol = /^[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\.[a-z]{2,5}(:[0-9]{1,5})?(\/.*)?$/
> 
>   if (validLink.match(validRegexWithoutProtocol)) validLink = `http://${url}/`
>   if (!validLink.match(validRegex)) validLink = `https://github.com/${validUsername}/`
> 
>   return validLink
> }
> ```
> 
> to
> 
> ```js
> function getUserLink(blog, githubProfileURL) {
>    const validRegexWithScheme = /same as now/
>    const validRegexWithoutScheme = /same as now/
>    if (validRegexWithSchema.test(blog)) return blog;
>    if (validRegexWithoutSchema.test(blog)) return `http://${blog}`;
>    return githubProfileURL
> }
> ```
> 
> Isn't this simpler?

@Roshanjossey I tried this earlier and 7 tests I made are failing.

Reply:
Did you change the params for calling the function?

Reply:
> Did you change the params for calling the function?

@Roshanjossey yes I did

Reply:
What's failing?

Reply:
FAIL  test/unit/helpers.test.js
  ● generateValidLink › returns valid link - valid URL format with `http` in between

    expect(received).toEqual(expected) // deep equality

    Expected: "http://tenshhttpiamd.com/"
    Received: "tenshhttpiamd.com"

      29 |     let validUrl = generateValidLink(url, username);
      30 |
    > 31 |     expect(validUrl).toEqual(`http://${url}/`);
         |                      ^
      32 |   });
      33 |
      34 |   test('returns valid link - valid URL format with `https` in between', async () => {

      at Object.<anonymous> (test/unit/helpers.test.js:31:22)

  ● generateValidLink › returns valid link - valid URL format with `https` in between

    expect(received).toEqual(expected) // deep equality

    Expected: "http://tenshhttpsiamd.com/"
    Received: "tenshhttpsiamd.com"

      36 |     let validUrl = generateValidLink(url, username);
      37 |
    > 38 |     expect(validUrl).toEqual(`http://${url}/`);
         |                      ^
      39 |   });
      40 |
      41 |   test('returns valid link - no protocol', async () => {

      at Object.<anonymous> (test/unit/helpers.test.js:38:22)

  ● generateValidLink › returns valid link - no protocol

    expect(received).toEqual(expected) // deep equality

    Expected: "http://tenshiamd.com/"
    Received: "tenshiamd.com"

      43 |     let validUrl = generateValidLink(url, username);
      44 |
    > 45 |     expect(validUrl).toEqual(`http://${url}/`);
         |                      ^
      46 |   });
      47 |
      48 |   test('returns valid link - no protocol and starting with `http`', async () => {

      at Object.<anonymous> (test/unit/helpers.test.js:45:22)

  ● generateValidLink › returns valid link - no protocol and starting with `http`

    expect(received).toEqual(expected) // deep equality

    Expected: "http://httptenshiamd.com/"
    Received: "httptenshiamd.com"

      50 |     let validUrl = generateValidLink(url, username);
      51 |
    > 52 |     expect(validUrl).toEqual(`http://${url}/`);
         |                      ^
      53 |   });
      54 |
      55 |   test('returns valid link - no protocol and starting with `https`', async () => {

      at Object.<anonymous> (test/unit/helpers.test.js:52:22)

  ● generateValidLink › returns valid link - no protocol and starting with `https`

    expect(received).toEqual(expected) // deep equality

    Expected: "http://httpstenshiamd.com/"
    Received: "httpstenshiamd.com"

      57 |     let validUrl = generateValidLink(url, username);
      58 |
    > 59 |     expect(validUrl).toEqual(`http://${url}/`);
         |                      ^
      60 |   });
      61 |
      62 |   test('returns valid link - incomplete URL format', async () => {

      at Object.<anonymous> (test/unit/helpers.test.js:59:22)

  ● generateValidLink › returns valid link - incomplete URL format

    expect(received).toEqual(expected) // deep equality

    Expected: "https://github.com/tenshiAMD/"
    Received: "tenshiAMD"

      64 |     let validUrl = generateValidLink(url, username);
      65 |
    > 66 |     expect(validUrl).toEqual(`https://github.com/${username}/`);
         |                      ^
      67 |   });
      68 |
      69 |   test('returns valid link - incomplete URL format with `null` username', async () => {

      at Object.<anonymous> (test/unit/helpers.test.js:66:22)

  ● generateValidLink › returns valid link - incomplete URL format with `null` username

    expect(received).toEqual(expected) // deep equality

    Expected: "https://github.com/all-contributors/"
    Received: null

      71 |     let validUrl = generateValidLink(url, null);
      72 |
    > 73 |     expect(validUrl).toEqual(`https://github.com/all-contributors/`);
         |                      ^
      74 |   });
      75 | });
      76 |

      at Object.<anonymous> (test/unit/helpers.test.js:73:22)

Reply:
> let validUrl = generateValidLink(url, username);

This is still the same parameters, the function I suggested takes in `html_url` as second param.

I think there's a communication problem here. Do you wanna get on a call and sort this out?
https://meet.jit.si/open

Reply:
@Roshanjossey hmm I see I still need to change my tests 😕 though it's the same implementation, this just takes my bandwidth. :(

Reply:
I could create a PR if you want

Reply:
> I could create a PR if you want

@Roshanjossey No need 😄 it will just take your time and I added some tweaks. LGTM now.